In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from itertools import cycle

pd.set_option('max_columns', 90)
pd.set_option('max_rows', 90)
plt.style.use('bmh')
color_pal = plt.rcParams['axes.prop_cycle'].by_key()['color']
color_cycle = cycle(plt.rcParams['axes.prop_cycle'].by_key()['color'])

## 1. Import the data

In [4]:
test_data=pd.read_csv("test.csv")
train_data=pd.read_csv("train.csv")

In [5]:
train_data.head()


UID  BLOCKID  SUMLEVEL  COUNTYID  STATEID        state state_ab  \
0  267822      NaN       140        53       36     New York       NY   
1  246444      NaN       140       141       18      Indiana       IN   
2  245683      NaN       140        63       18      Indiana       IN   
3  279653      NaN       140       127       72  Puerto Rico       PR   
4  247218      NaN       140       161       20       Kansas       KS   

         city           place   type primary  zip_code  area_code        lat  \
0    Hamilton        Hamilton   City   tract     13346        315  42.840812   
1  South Bend        Roseland   City   tract     46616        574  41.701441   
2    Danville        Danville   City   tract     46122        317  39.792202   
3    San Juan        Guaynabo  Urban   tract       927        787  18.396103   
4   Manhattan  Manhattan City   City   tract     66502        785  39.195573   

         lng        ALand   AWater   pop  male_pop  female_pop  rent_mean  \
0 -75.501524  202183361.0  1699120  5230      2612        2618  769.38638   
1 -86.266614    1560828.0   100363  2633      1349        1284  804.87924   
2 -86.515246   69561595.0   284193  6881      3643        3238  742.77365   
3 -66.104169    1105793.0        0  2700      1141        1559  803.42018   
4 -96.569366    2554403.0        0  5637      2586        3051  938.56493   

   rent_median  rent_stdev  rent_sample_weight  rent_samples  rent_gt_10  \
0        784.0   232.63967           272.34441         362.0     0.86761   
1        848.0   253.46747           312.58622         513.0     0.97410   
2        703.0   323.39011           291.85520         378.0     0.95238   
3        782.0   297.39258           259.30316         368.0     0.94693   
4        881.0   392.44096          1005.42886        1704.0     0.99286   

   rent_gt_15  rent_gt_20  rent_gt_25  rent_gt_30  rent_gt_35  rent_gt_40  \
0     0.79155     0.59155     0.45634     0.42817     0.18592     0.15493   
1     0.93227     0.69920     0.69920     0.55179     0.41235     0.39044   
2     0.88624     0.79630     0.66667     0.39153     0.39153     0.28307   
3     0.87151     0.69832     0.61732     0.51397     0.46927     0.35754   
4     0.98247     0.91688     0.84740     0.78247     0.60974     0.55455   

   rent_gt_50  universe_samples  used_samples      hi_mean  hi_median  \
0     0.12958               387           355  63125.28406    48120.0   
1     0.27888               542           502  41931.92593    35186.0   
2     0.15873               459           378  84942.68317    74964.0   
3     0.32961               438           358  48733.67116    37845.0   
4     0.44416              1725          1540  31834.15466    22497.0   

      hi_stdev  hi_sample_weight  hi_samples  family_mean  family_median  \
0  49042.01206        1290.96240      2024.0  67994.14790        53245.0   
1  31639.50203         838.74664      1127.0  50670.10337        43023.0   
2  56811.62186        1155.20980      2488.0  95262.51431        85395.0   
3  45100.54010         928.32193      1267.0  56401.68133        44399.0   
4  34046.50907        1548.67477      1983.0  54053.42396        50272.0   

   family_stdev  family_sample_weight  family_samples  hc_mortgage_mean  \
0   47667.30119             884.33516          1491.0        1414.80295   
1   34715.57548             375.28798           554.0         864.41390   
2   49292.67664             709.74925          1889.0        1506.06758   
3   41082.90515             490.18479           729.0        1175.28642   
4   39609.12605             244.08903           395.0        1192.58759   

   hc_mortgage_median  hc_mortgage_stdev  hc_mortgage_sample_weight  \
0              1223.0          641.22898                  377.83135   
1               784.0          482.27020                  316.88320   
2              1361.0          731.89394                  699.41354   
3              1101.0          428.98751                  261.28471   
4

In [6]:
test_data.head()

UID  BLOCKID  SUMLEVEL  COUNTYID  STATEID         state state_ab  \
0  255504      NaN       140       163       26      Michigan       MI   
1  252676      NaN       140         1       23         Maine       ME   
2  276314      NaN       140        15       42  Pennsylvania       PA   
3  248614      NaN       140       231       21      Kentucky       KY   
4  286865      NaN       140       355       48         Texas       TX   

             city                  place     type primary  zip_code  \
0         Detroit  Dearborn Heights City      CDP   tract     48239   
1          Auburn            Auburn City     City   tract      4210   
2       Pine City              Millerton  Borough   tract     14871   
3      Monticello        Monticello City     City   tract     42633   
4  Corpus Christi                  Edroy     Town   tract     78410   

   area_code        lat        lng      ALand   AWater   pop  male_pop  \
0        313  42.346422 -83.252823    2711280    39555  3417      1479   
1        207  44.100724 -70.257832   14778785  2705204  3796      1846   
2        607  41.948556 -76.783808  258903666   863840  3944      2065   
3        606  36.746009 -84.766870  501694825  2623067  2508      1427   
4        361  27.882461 -97.678586   13796057   497689  6230      3274   

   female_pop   rent_mean  rent_median  rent_stdev  rent_sample_weight  \
0        1938   858.57169        859.0   232.39082           276.07497   
1        1950   832.68625        750.0   267.22342           183.32299   
2        1879   816.00639        755.0   416.25699           141.39063   
3        1081   418.68937        385.0   156.92024            88.95960   
4        2956  1031.63763        997.0   326.76727           277.39844   

   rent_samples  rent_gt_10  rent_gt_15  rent_gt_20  rent_gt_25  rent_gt_30  \
0         424.0     1.00000     0.95696     0.85316     0.85316     0.85316   
1         245.0     1.00000     1.00000     0.86611     0.67364     0.30962   
2         217.0     0.97573     0.93204     0.78641     0.71845     0.63592   
3          93.0     1.00000     0.93548     0.93548     0.64516     0.55914   
4         624.0     0.72276     0.66506     0.53526     0.38301     0.18910   

   rent_gt_35  rent_gt_40  rent_gt_50  universe_samples  used_samples  \
0     0.85316     0.76962     0.63544               435           395   
1     0.30962     0.30962     0.27197               275           239   
2     0.47573     0.43689     0.32524               245           206   
3     0.46237     0.46237     0.36559               153            93   
4     0.16667     0.14263     0.11058               660           624   

        hi_mean  hi_median     hi_stdev  hi_sample_weight  hi_samples  \
0   48899.52121    38746.0  44392.20902         798.02401      1180.0   
1   72335.33234    61008.0  51895.81159         922.82969      1722.0   
2   58501.15901    51648.0  45245.27248         893.07759      1461.0   
3   38237.55059    31612.0  34527.61607         775.17947       957.0   
4  114456.07790    94211.0  81950.95692         836.30759      2404.0   

    family_mean  family_median  family_stdev  family_sample_weight  \
0   53802.87122        45167.0   43756.56479             464.30972   
1   85642.22095        74759.0   49156.72870             482.99945   
2   65694.06582        57186.0   44239.31893             619.73962   
3   44156.38709        34687.0   34899.74300             535.21987   
4  123527.02420       103898.0   72173.55823             507.42257   

   family_samples  hc_mortgage_mean  hc_mortgage_median  hc_mortgage_stdev  \
0           769.0        1139.24548              1109.0          336.47710   
1          1147.0        1533.25988              1438.0          536.61118   
2          1084.0        1254.54462              1089.0          596.85204   
3           689.0         862.65763               749.0          624.42157   
4          1738.0        1996.41425              1907.0          740.21168   

 

## 2. Find the primary key

UID is the primary key, as it is unique reference key.

## 3.Gauge the fill rate of the variables and devise plans for missing value treatment. Please explain explicitly the reason for the treatment chosen for each variable. 

First lets perform operations on the training data, then perform on the test data.

In [7]:
train_data.dtypes

UID                              int64
BLOCKID                        float64
SUMLEVEL                         int64
COUNTYID                         int64
STATEID                          int64
state                           object
state_ab                        object
city                            object
place                           object
type                            object
primary                         object
zip_code                         int64
area_code                        int64
lat                            float64
lng                            float64
ALand                          float64
AWater                           int64
pop                              int64
male_pop                         int64
female_pop                       int64
rent_mean                      float64
rent_median                    float64
rent_stdev                     float64
rent_sample_weight             float64
rent_samples                   float64
rent_gt_10               

In [8]:
train_data.columns

Index(['UID', 'BLOCKID', 'SUMLEVEL', 'COUNTYID', 'STATEID', 'state',
       'state_ab', 'city', 'place', 'type', 'primary', 'zip_code', 'area_code',
       'lat', 'lng', 'ALand', 'AWater', 'pop', 'male_pop', 'female_pop',
       'rent_mean', 'rent_median', 'rent_stdev', 'rent_sample_weight',
       'rent_samples', 'rent_gt_10', 'rent_gt_15', 'rent_gt_20', 'rent_gt_25',
       'rent_gt_30', 'rent_gt_35', 'rent_gt_40', 'rent_gt_50',
       'universe_samples', 'used_samples', 'hi_mean', 'hi_median', 'hi_stdev',
       'hi_sample_weight', 'hi_samples', 'family_mean', 'family_median',
       'family_stdev', 'family_sample_weight', 'family_samples',
       'hc_mortgage_mean', 'hc_mortgage_median', 'hc_mortgage_stdev',
       'hc_mortgage_sample_weight', 'hc_mortgage_samples', 'hc_mean',
       'hc_median', 'hc_stdev', 'hc_samples', 'hc_sample_weight',
       'home_equity_second_mortgage', 'second_mortgage', 'home_equity', 'debt',
       'second_mortgage_cdf', 'home_equity_cdf', 'debt_cdf', '

In [9]:
A=pd.DataFrame(train_data.columns)
A

0
0                           UID
1                       BLOCKID
2                      SUMLEVEL
3                      COUNTYID
4                       STATEID
5                         state
6                      state_ab
7                          city
8                         place
9                          type
10                      primary
11                     zip_code
12                    area_code
13                          lat
14                          lng
15                        ALand
16                       AWater
17                          pop
18                     male_pop
19                   female_pop
20                    rent_mean
21                  rent_median
22                   rent_stdev
23           rent_sample_weight
24                 rent_samples
25                   rent_gt_10
26                   rent_gt_15
27                   rent_gt_20
28                   rent_gt_25
29                   rent_gt_30
30                   rent_gt_35
31                   rent_gt_40
32                   rent_gt_50
33             universe_samples
34                 used_samples
35                      hi_mean
36                    hi_median
37                     hi_stdev
38             hi_sample_weight
39                   hi_samples
40                  family_mean
41                family_median
42                 family_stdev
43         family_sample_weight
44               family_samples
45             hc_mortgage_mean
46           hc_mortgage_median
47            hc_mortgage_stdev
48    hc_mortgage_sample_weight
49          hc_mortgage_samples
50                      hc_mean
51                    hc_median
52                     hc_stdev
53                   hc_samples
54             hc_sample_weight
55  home_equity_second_mortgage
56              second_mortgage
57                  home_equity
58                         debt
59          second_mortgage_cdf
60              home_equity_cdf
61                     debt_cdf
62                    hs_degree
63               hs_degree_male
64             hs_degree_female
65                male_age_mean
66              male_age_median
67               male_age_stdev
68       male_age_sample_weight
69             male_age_samples
70              female_age_mean
71            female_age_median
72             female_age_stdev
73     female_age_sample_weight
74           female_age_samples
75                      pct_own
76                      married
77                  married_snp
78                    separated
79                     divorced

In [10]:
train_data.shape

(27321, 80)

In [11]:
train_data.isnull().sum()

UID                                0
BLOCKID                        27321
SUMLEVEL                           0
COUNTYID                           0
STATEID                            0
state                              0
state_ab                           0
city                               0
place                              0
type                               0
primary                            0
zip_code                           0
area_code                          0
lat                                0
lng                                0
ALand                              0
AWater                             0
pop                                0
male_pop                           0
female_pop                         0
rent_mean                        314
rent_median                      314
rent_stdev                       314
rent_sample_weight               314
rent_samples                     314
rent_gt_10                       314
rent_gt_15                       314
r

In [12]:
train_data.duplicated().sum()

160

1.There are 80 columns with 27321 records in the training data </br>
2.There are null values in the columns</br>
3.There are 160 duplicated records</br>
4.The data types range from int to float 

## Treating Missing Values

## Number of columns with missing values

In [13]:
len(train_data.columns[train_data.isnull().sum(axis=0 )> 0])

59

There are 59 columns with missing values.

## percentage of missings values in each column 

In [14]:
percent_missing = train_data.isnull().sum() * 100 / len(train_data)
missing_value_df = pd.DataFrame({'column_name': train_data.columns,
                                 'percent_missing': percent_missing})

In [15]:
missing_value_df

column_name  percent_missing
UID                                                  UID         0.000000
BLOCKID                                          BLOCKID       100.000000
SUMLEVEL                                        SUMLEVEL         0.000000
COUNTYID                                        COUNTYID         0.000000
STATEID                                          STATEID         0.000000
state                                              state         0.000000
state_ab                                        state_ab         0.000000
city                                                city         0.000000
place                                              place         0.000000
type                                                type         0.000000
primary                                          primary         0.000000
zip_code                                        zip_code         0.000000
area_code                                      area_code         0.000000
lat                                                  lat         0.000000
lng                                                  lng         0.000000
ALand                                              ALand         0.000000
AWater                                            AWater         0.000000
pop                                                  pop         0.000000
male_pop                                        male_pop         0.000000
female_pop                                    female_pop         0.000000
rent_mean                                      rent_mean         1.149299
rent_median                                  rent_median         1.149299
rent_stdev                                    rent_stdev         1.149299
rent_sample_weight                    rent_sample_weight         1.149299
rent_samples                                rent_samples         1.149299
rent_gt_10                                    rent_gt_10         1.149299
rent_gt_15                                    rent_gt_15         1.149299
rent_gt_20                                    rent_gt_20         1.149299
rent_gt_25                                    rent_gt_25         1.149299
rent_gt_30                                    rent_gt_30         1.149299
rent_gt_35                                    rent_gt_35         1.149299
rent_gt_40                                    rent_gt_40         1.149299
rent_gt_50                                    rent_gt_50         1.149299
universe_samples                        universe_samples         0.000000
used_samples                                used_samples         0.000000
hi_mean                                          hi_mean         0.980930
hi_median                                      hi_median         0.980930
hi_stdev                                        hi_stdev         0.980930
hi_sample_weight                        hi_sample_weight         0.980930
hi_samples                                    hi_samples         0.980930
family_mean                                  family_mean         1.090736
family_median                              family_median         1.090736
family_stdev                                family_stdev         1.090736
family_sample_weight                family_sample_weight         1.090736
family_samples                            family_samples         1.090736
hc_mortgage_mean                        hc_mortgage_mean         2.097288
hc_mortgage_median                    hc_mortgage_median         2.097288
hc_mortgage_stdev                      hc_mortgage_stdev         2.097288
hc_mortgage_sample_weight      hc_mortgage_sample_weight         2.097288
hc_mortgage_samples                  hc_mortgage_samples         2.097288
hc_mean                                          hc_mean         2.196113
hc_median                                      hc_median         2.196113
hc_stdev                                        hc_stdev         2.196113
hc_samples                                    hc_


We can delete the column "BLOCK ID" as it has all null values 

In [16]:
train_data.drop('BLOCKID', axis=1 , inplace=True)

## Calculate the overall percentage of missing data in the dataset

In [17]:
Missing_values=train_data[train_data.isnull().any(axis=1)]
Missing_values

UID  SUMLEVEL  COUNTYID  STATEID          state state_ab  \
51     223593       140        19        4        Arizona       AZ   
94     233040       140       101        8       Colorado       CO   
153    263292       140        13       34     New Jersey       NJ   
302    267158       140        47       36       New York       NY   
340    292484       140        25       55      Wisconsin       WI   
...       ...       ...       ...      ...            ...      ...   
27127  266321       140         5       36       New York       NY   
27175  235725       140        57       12        Florida       FL   
27176  247777       140        61       21       Kentucky       KY   
27216  266166       140         5       36       New York       NY   
27240  251078       140        25       25  Massachusetts       MA   

              city              place  type primary  zip_code  area_code  \
51          Tucson         Littletown   CDP   tract     85734        520   
94          Pueblo        Pueblo City  City   tract     81001        719   
153         Newark        Silver Lake  City   tract      7107        973   
302       Brooklyn      New York City  City   tract     11215        718   
340        Madison       Madison City  City   tract     53703        608   
...            ...                ...   ...     ...       ...        ...   
27127        Bronx  Mount Vernon City  City   tract     10458        718   
27175        Tampa       Pebble Creek  City   tract     33647        813   
27176  Brownsville   Brownsville City  City   tract     42210        270   
27216        Bronx       Pelham Manor  City   tract     10462        718   
27240       Boston          Brookline  City   tract      2124        617   

             lat         lng        ALand   AWater   pop  male_pop  \
51     32.067721 -110.867177    2909152.0        0  4531      4370   
94     38.308407 -104.577687    1042529.0        0   579       270   
153    40.773830  -74.175141     763973.0    22300  3458      1787   
302    40.659126  -73.969773    2313042.0   227326     0         0   
340    43.073403  -89.395430     175748.0        0  3274      1293   
...          ...         ...          ...      ...   ...       ...   
27127  40.874095  -73.885197     173358.0        0  7561      3640   
27175  28.149447  -82.356517    2464707.0    50567  1105       457   
27176  37.197158  -86.156329  175916489.0  2094374    17        11   
27216  40.856476  -73.865431     124862.0        0  3502      1441   
27240  42.300613  -71.094613    2076316.0    27539   360       264   

       female_pop   rent_mean  rent_median  rent_stdev  rent_sample_weight  \
51            161         NaN          NaN         NaN                 NaN   
94            309   782.00000        781.0    22.95830             4.00000   
153          1671   890.69365        929.0   472.84907           566.54167   
302             0         NaN          NaN         NaN                 NaN   
340          1981  1191.78679        956.0   737.36689           768.38091   
...           ...         ...          ...         ...                 ...   
27127        3921  1183.94615       1171.0   410.85034           877.82915   
27175         648  1144.84723       1135.0   204.04871           133.90349   
27176           6         NaN          NaN         NaN                 NaN   
27216        2061  1302.60910       1270.0   380.82042           295.19409   
27240          96         NaN          NaN         NaN                 NaN   

       rent_samples  rent_gt_10  rent_gt_15  rent_gt_20  rent_gt_25  \
51              NaN         NaN         NaN         NaN         NaN   
94              4.0     1.00000     1.00000     1.00000     1.00000   
153          1011.0     0.96211     0.79163     0.72981     0.66899   
302             NaN         NaN         NaN         NaN         NaN   
340          1497.0     1.00000     0.98039     0.96757     0.96757   
...             ...         ...         ...         

In [18]:
Missing_values.shape

(736, 79)

In [19]:
train_data.shape

(27321, 79)

In [20]:
missing_percentage= (736/27321) * 100
missing_percentage

2.693898466381172

The overall missing data in the dataset is 2.69 which is negligable, therefore the missing values can be dropped.

## Dropping NULL/NAN values 

In [21]:
train_data.dropna()

UID  SUMLEVEL  COUNTYID  STATEID         state state_ab  \
0      267822       140        53       36      New York       NY   
1      246444       140       141       18       Indiana       IN   
2      245683       140        63       18       Indiana       IN   
3      279653       140       127       72   Puerto Rico       PR   
4      247218       140       161       20        Kansas       KS   
...       ...       ...       ...      ...           ...      ...   
27316  279212       140        43       72   Puerto Rico       PR   
27317  277856       140        91       42  Pennsylvania       PA   
27318  233000       140        87        8      Colorado       CO   
27319  287425       140       439       48         Texas       TX   
27320  265371       140         3       32        Nevada       NV   

              city             place     type primary  zip_code  area_code  \
0         Hamilton          Hamilton     City   tract     13346        315   
1       South Bend          Roseland     City   tract     46616        574   
2         Danville          Danville     City   tract     46122        317   
3         San Juan          Guaynabo    Urban   tract       927        787   
4        Manhattan    Manhattan City     City   tract     66502        785   
...            ...               ...      ...     ...       ...        ...   
27316        Coamo             Coamo    Urban   tract       769        787   
27317    Blue Bell         Blue Bell  Borough   tract     19422        215   
27318      Weldona      Saddle Ridge     City   tract     80653        970   
27319  Colleyville  Colleyville City     Town   tract     76034        817   
27320    Las Vegas          Paradise     City   tract     89123        702   

             lat         lng         ALand    AWater    pop  male_pop  \
0      42.840812  -75.501524  2.021834e+08   1699120   5230      2612   
1      41.701441  -86.266614  1.560828e+06    100363   2633      1349   
2      39.792202  -86.515246  6.956160e+07    284193   6881      3643   
3      18.396103  -66.104169  1.105793e+06         0   2700      1141   
4      39.195573  -96.569366  2.554403e+06         0   5637      2586   
...          ...         ...           ...       ...    ...       ...   
27316  18.076060  -66.358379  6.970300e+05         0   1847       909   
27317  40.158138  -75.307271  5.077337e+06     11786   4155      2116   
27318  40.410316 -103.814003  1.323262e+09  17577610   2829      1465   
27319  32.904866  -97.162151  1.865230e+07    158882  11542      5727   
27320  36.064754 -115.152237  7.796308e+06         0   3726      1815   

       female_pop   rent_mean  rent_median  rent_stdev  rent_sample_weight  \
0            2618   769.38638        784.0   232.63967           272.34441   
1            1284   804.87924        848.0   253.46747           312.58622   
2            3238   742.77365        703.0   323.39011           291.85520   
3            1559   803.42018        782.0   297.39258           259.30316   
4            3051   938.56493        881.0   392.44096          1005.42886   
...           ...         ...          ...         ...                 ...   
27316         938   439.42839        419.0   140.29970           170.00000   
27317        2039  1813.19253       1788.0   492.92300            64.84927   
27318        1364   849.39107        834.0   336.47530           120.91448   
27319        5815  1972.45746       1843.0   633.02173            19.16328   
27320        1911   949.84199        924.0   198.82109           555.87526   

       rent_samples  rent_gt_10  rent_gt_15  rent_gt_20  rent_gt_25  \
0             362.0     0.86761     0.79155     0.59155     0.45634   
1             513.0     0.97410     0.93227     0.69920     0.69920   
2             378.0     0.95238     0.88624     0.79630     0.66667   
3             368.0     0.94693     0.87151     0.69832     0.61732   
4            1704.0     0.99286     0.98247     0.91688     0.84740   
... 

In [22]:
train_data.duplicated().sum()

160

In [23]:
train_data.drop_duplicates(inplace=True)

In [24]:
train_data.duplicated().sum()

0

## Final dataset size

In [25]:
train_data.shape

(27161, 79)

## Same will be performed on the test data

In [26]:
test_data.isnull().sum()

UID                                0
BLOCKID                        11709
SUMLEVEL                           0
COUNTYID                           0
STATEID                            0
state                              0
state_ab                           0
city                               0
place                              0
type                               0
primary                            0
zip_code                           0
area_code                          0
lat                                0
lng                                0
ALand                              0
AWater                             0
pop                                0
male_pop                           0
female_pop                         0
rent_mean                        148
rent_median                      148
rent_stdev                       148
rent_sample_weight               148
rent_samples                     148
rent_gt_10                       149
rent_gt_15                       149
r

In [27]:
percent_missing1 = test_data.isnull().sum() * 100 / len(test_data)
missing_value_df1 = pd.DataFrame({'column_name': test_data.columns,
                                 'percent_missing': percent_missing})

In [28]:
missing_value_df1

column_name  percent_missing
UID                                                  UID         0.000000
BLOCKID                                          BLOCKID       100.000000
SUMLEVEL                                        SUMLEVEL         0.000000
COUNTYID                                        COUNTYID         0.000000
STATEID                                          STATEID         0.000000
state                                              state         0.000000
state_ab                                        state_ab         0.000000
city                                                city         0.000000
place                                              place         0.000000
type                                                type         0.000000
primary                                          primary         0.000000
zip_code                                        zip_code         0.000000
area_code                                      area_code         0.000000
lat                                                  lat         0.000000
lng                                                  lng         0.000000
ALand                                              ALand         0.000000
AWater                                            AWater         0.000000
pop                                                  pop         0.000000
male_pop                                        male_pop         0.000000
female_pop                                    female_pop         0.000000
rent_mean                                      rent_mean         1.149299
rent_median                                  rent_median         1.149299
rent_stdev                                    rent_stdev         1.149299
rent_sample_weight                    rent_sample_weight         1.149299
rent_samples                                rent_samples         1.149299
rent_gt_10                                    rent_gt_10         1.149299
rent_gt_15                                    rent_gt_15         1.149299
rent_gt_20                                    rent_gt_20         1.149299
rent_gt_25                                    rent_gt_25         1.149299
rent_gt_30                                    rent_gt_30         1.149299
rent_gt_35                                    rent_gt_35         1.149299
rent_gt_40                                    rent_gt_40         1.149299
rent_gt_50                                    rent_gt_50         1.149299
universe_samples                        universe_samples         0.000000
used_samples                                used_samples         0.000000
hi_mean                                          hi_mean         0.980930
hi_median                                      hi_median         0.980930
hi_stdev                                        hi_stdev         0.980930
hi_sample_weight                        hi_sample_weight         0.980930
hi_samples                                    hi_samples         0.980930
family_mean                                  family_mean         1.090736
family_median                              family_median         1.090736
family_stdev                                family_stdev         1.090736
family_sample_weight                family_sample_weight         1.090736
family_samples                            family_samples         1.090736
hc_mortgage_mean                        hc_mortgage_mean         2.097288
hc_mortgage_median                    hc_mortgage_median         2.097288
hc_mortgage_stdev                      hc_mortgage_stdev         2.097288
hc_mortgage_sample_weight      hc_mortgage_sample_weight         2.097288
hc_mortgage_samples                  hc_mortgage_samples         2.097288
hc_mean                                          hc_mean         2.196113
hc_median                                      hc_median         2.196113
hc_stdev                                        hc_stdev         2.196113
hc_samples                                    hc_

In [29]:
test_data.drop('BLOCKID', axis=1 , inplace=True)

In [30]:
Missing_values1=test_data[test_data.isnull().any(axis=1)]
Missing_values1

UID  SUMLEVEL  COUNTYID  STATEID          state state_ab  \
17     265339       140         3       32         Nevada       NV   
27     287596       140       451       48          Texas       TX   
44     250903       140        25       25  Massachusetts       MA   
54     287557       140       441       48          Texas       TX   
70     247510       140       209       20         Kansas       KS   
...       ...       ...       ...      ...            ...      ...   
11543  267675       140        47       36       New York       NY   
11554  288512       140        35       49           Utah       UT   
11561  275715       140         3       42   Pennsylvania       PA   
11636  276588       140        27       42   Pennsylvania       PA   
11675  266164       140         5       36       New York       NY   

                 city                place     type primary  zip_code  \
17          Las Vegas           Winchester     City   tract     89119   
27         San Angelo      San Angelo City     Town   tract     76903   
44          Cambridge       Cambridge City     City   tract      2139   
54            Abilene             Tye City     Town   tract     79607   
70        Kansas City     Kansas City City     City   tract     66104   
...               ...                  ...      ...     ...       ...   
11543        Brooklyn        New York City     City   tract     11212   
11554  Salt Lake City  Salt Lake City City     City   tract     84116   
11561      Pittsburgh      Pittsburgh City  Borough   tract     15219   
11636   State College        State College  Borough   tract     16801   
11675           Bronx    Mount Vernon City     City   tract     10452   

       area_code        lat         lng     ALand   AWater   pop  male_pop  \
17           702  36.111448 -115.159125    612803        0   724       490   
27           325  31.431831 -100.404539   4603626        0  3099      2231   
44           617  42.359478  -71.121263   1158117        0  6234      2518   
54           325  32.423876  -99.842475  18425872   194540   916       857   
70           913  39.171767  -94.680617   8005557  1678971     0         0   
...          ...        ...         ...       ...      ...   ...       ...   
11543        718  40.670130  -73.908859    195831        0  4318      1772   
11554        801  40.788810 -111.992473  39295590    83499     0         0   
11561        412  40.441658  -79.976462    183645        0   327       150   
11636        814  40.801303  -77.856566    931782        0  6492      2903   
11675        718  40.840525  -73.915088     92765        0  5108      2565   

       female_pop   rent_mean  rent_median  rent_stdev  rent_sample_weight  \
17            234   751.50703        736.0   172.34699           446.71620   
27            868  1061.03134       1079.0   221.25258            32.20117   
44           3716  1964.19623       2025.0   769.92869           137.01400   
54             59   972.00000        968.0    50.76498             2.96970   
70              0         NaN          NaN         NaN                 NaN   
...           ...         ...          ...         ...                 ...   
11543        2546   610.68727        434.0   506.85855          1313.88139   
11554           0         NaN          NaN         NaN                 NaN   
11561         177   234.28539        206.0   193.59771           137.26000   
11636        3589  1512.16288       1661.0   647.45329            29.14141   
11675        2543  1027.02034       1072.0   487.29451           830.19199   

       rent_samples  rent_gt_10  rent_gt_15  rent_gt_20  rent_gt_25  \
17            501.0     0.95781     0.91772     0.87764     0.77637   
27             91.0     1.00000     1.00000     1.00000     0.89011   
44            646.0     1.00000     0.98262     0.90521     0.87204   
54              6.0     1.00000     1.00000     1.00000     1.00000   
70              NaN         NaN         NaN         NaN        

In [31]:
test_data.shape

(11709, 79)

In [32]:
missing_percentage= (354/11709) * 100
missing_percentage

3.0233153984114787

In [33]:
test_data.dropna()

UID  SUMLEVEL  COUNTYID  STATEID          state state_ab  \
0      255504       140       163       26       Michigan       MI   
1      252676       140         1       23          Maine       ME   
2      276314       140        15       42   Pennsylvania       PA   
3      248614       140       231       21       Kentucky       KY   
4      286865       140       355       48          Texas       TX   
...       ...       ...       ...      ...            ...      ...   
11704  238088       140       105       12        Florida       FL   
11705  242811       140        31       17       Illinois       IL   
11706  250127       140         9       25  Massachusetts       MA   
11707  241096       140        27       19           Iowa       IA   
11708  287763       140       453       48          Texas       TX   

                 city                  place     type primary  zip_code  \
0             Detroit  Dearborn Heights City      CDP   tract     48239   
1              Auburn            Auburn City     City   tract      4210   
2           Pine City              Millerton  Borough   tract     14871   
3          Monticello        Monticello City     City   tract     42633   
4      Corpus Christi                  Edroy     Town   tract     78410   
...               ...                    ...      ...     ...       ...   
11704        Lakeland        Crystal Springs     City   tract     33810   
11705         Chicago           Chicago City  Village   tract     60609   
11706        Lawrence      Methuen Town City     City   tract      1841   
11707         Carroll           Carroll City     City   tract     51401   
11708          Austin     Sunset Valley City     Town   tract     78745   

       area_code        lat        lng      ALand   AWater   pop  male_pop  \
0            313  42.346422 -83.252823    2711280    39555  3417      1479   
1            207  44.100724 -70.257832   14778785  2705204  3796      1846   
2            607  41.948556 -76.783808  258903666   863840  3944      2065   
3            606  36.746009 -84.766870  501694825  2623067  2508      1427   
4            361  27.882461 -97.678586   13796057   497689  6230      3274   
...          ...        ...        ...        ...      ...   ...       ...   
11704        863  28.226068 -82.068886   92582775  1166617  5611      2697   
11705        773  41.804936 -87.667304     327029        0  2695      1504   
11706        978  42.737778 -71.131761    5225804   393810  7392      3669   
11707        712  42.081366 -94.866175   11066759        0  5945      2732   
11708        512  30.219013 -97.774728    1990126        0  4117      2070   

       female_pop   rent_mean  rent_median  rent_stdev  rent_sample_weight  \
0            1938   858.57169        859.0   232.39082           276.07497   
1            1950   832.68625        750.0   267.22342           183.32299   
2            1879   816.00639        755.0   416.25699           141.39063   
3            1081   418.68937        385.0   156.92024            88.95960   
4            2956  1031.63763        997.0   326.76727           277.39844   
...           ...         ...          ...         ...                 ...   
11704        2914  1458.82449       1603.0   566.90682            29.43733   
11705        1191   700.53513        661.0   254.66700           480.86455   
11706        3723  1069.70567       1138.0   488.13975           207.29615   
11707        3213   696.93368        576.0   595.16228           503.83775   
11708        2047   950.09294        864.0   333.82364           417.07457   

       rent_samples  rent_gt_10  rent_gt_15  rent_gt_20  rent_gt_25  \
0             424.0     1.00000     0.95696     0.85316     0.85316   
1             245.0     1.00000     1.00000     0.86611     0.67364   
2             217.0     0.97573     0.93204     0.78641     0.71845   
3              93.0     1.00000     0.93548     0.93548     0.64516   
4             624.0     0.72276     0.6

In [34]:
test_data.drop_duplicates(inplace=True)

In [35]:
test_data.duplicated().sum()

0

## TEST DATA FINAL SIZE 

In [36]:
test_data.shape

(11677, 79)

### Perform EDA and come out with insights into population density and age. You may require deriving new fields (Make sure to weight averages for accurate measurements): 
### •	Population density (hint-use ‘pop’ and ‘Aland’ to calculate)</br>
### •	median age (hint-use the variables ‘male_age_median’, ‘female_age_median’, ‘male_pop’, ‘female_pop’)</br>
### Visualize the findings using appropriate chart type.


In [37]:
df= train_data[['state', 'city', 'place','type','ALand', 'pop', 'male_age_median', 'female_age_median', 'male_pop', 'female_pop']]
df.head()

state        city           place   type        ALand   pop  \
0     New York    Hamilton        Hamilton   City  202183361.0  5230   
1      Indiana  South Bend        Roseland   City    1560828.0  2633   
2      Indiana    Danville        Danville   City   69561595.0  6881   
3  Puerto Rico    San Juan        Guaynabo  Urban    1105793.0  2700   
4       Kansas   Manhattan  Manhattan City   City    2554403.0  5637   

   male_age_median  female_age_median  male_pop  female_pop  
0         44.00000           45.33333      2612        2618  
1         32.00000           37.58333      1349        1284  
2         40.83333           42.83333      3643        3238  
3         48.91667           50.58333      1141        1559  
4         22.41667           21.58333      2586        3051

In [38]:
df.shape

(27161, 10)

## Population density is total population by total area of land; Therefore we will use total population as pop and total area of land as ALand 

In [39]:
pd.options.mode.chained_assignment = None 

In [40]:
df['p_density'] = df['pop'] / df['ALand']

## Median age will be average of of median age of male and median age of female.

In [45]:
df['median_age'] = (df['male_age_median']+ df['female_age_median']) / 2

In [46]:
df.head()

state        city           place   type        ALand   pop  \
0     New York    Hamilton        Hamilton   City  202183361.0  5230   
1      Indiana  South Bend        Roseland   City    1560828.0  2633   
2      Indiana    Danville        Danville   City   69561595.0  6881   
3  Puerto Rico    San Juan        Guaynabo  Urban    1105793.0  2700   
4       Kansas   Manhattan  Manhattan City   City    2554403.0  5637   

   male_age_median  female_age_median  male_pop  female_pop  p_density  \
0         44.00000           45.33333      2612        2618   0.000026   
1         32.00000           37.58333      1349        1284   0.001687   
2         40.83333           42.83333      3643        3238   0.000099   
3         48.91667           50.58333      1141        1559   0.002442   
4         22.41667           21.58333      2586        3051   0.002207   

   median_age  
0   44.666665  
1   34.791665  
2   41.833330  
3   49.750000  
4   22.000000

Load the df dataset in the excel sheet to perform visualisations.

In [49]:
df.to_excel('A.xlsx')

In [69]:
#--> USED POWERBI For visualisations


### Create bins for population into a new variable by selecting appropriate class interval so that the no of categories(bins) don’t exceed 5 for the ease of analysis. Analyze the married, separated and divorced population for these population brackets. Visualize using appropriate chart type.


In [52]:
train_data.head()

UID  SUMLEVEL  COUNTYID  STATEID        state state_ab        city  \
0  267822       140        53       36     New York       NY    Hamilton   
1  246444       140       141       18      Indiana       IN  South Bend   
2  245683       140        63       18      Indiana       IN    Danville   
3  279653       140       127       72  Puerto Rico       PR    San Juan   
4  247218       140       161       20       Kansas       KS   Manhattan   

            place   type primary  zip_code  area_code        lat        lng  \
0        Hamilton   City   tract     13346        315  42.840812 -75.501524   
1        Roseland   City   tract     46616        574  41.701441 -86.266614   
2        Danville   City   tract     46122        317  39.792202 -86.515246   
3        Guaynabo  Urban   tract       927        787  18.396103 -66.104169   
4  Manhattan City   City   tract     66502        785  39.195573 -96.569366   

         ALand   AWater   pop  male_pop  female_pop  rent_mean  rent_median  \
0  202183361.0  1699120  5230      2612        2618  769.38638        784.0   
1    1560828.0   100363  2633      1349        1284  804.87924        848.0   
2   69561595.0   284193  6881      3643        3238  742.77365        703.0   
3    1105793.0        0  2700      1141        1559  803.42018        782.0   
4    2554403.0        0  5637      2586        3051  938.56493        881.0   

   rent_stdev  rent_sample_weight  rent_samples  rent_gt_10  rent_gt_15  \
0   232.63967           272.34441         362.0     0.86761     0.79155   
1   253.46747           312.58622         513.0     0.97410     0.93227   
2   323.39011           291.85520         378.0     0.95238     0.88624   
3   297.39258           259.30316         368.0     0.94693     0.87151   
4   392.44096          1005.42886        1704.0     0.99286     0.98247   

   rent_gt_20  rent_gt_25  rent_gt_30  rent_gt_35  rent_gt_40  rent_gt_50  \
0     0.59155     0.45634     0.42817     0.18592     0.15493     0.12958   
1     0.69920     0.69920     0.55179     0.41235     0.39044     0.27888   
2     0.79630     0.66667     0.39153     0.39153     0.28307     0.15873   
3     0.69832     0.61732     0.51397     0.46927     0.35754     0.32961   
4     0.91688     0.84740     0.78247     0.60974     0.55455     0.44416   

   universe_samples  used_samples      hi_mean  hi_median     hi_stdev  \
0               387           355  63125.28406    48120.0  49042.01206   
1               542           502  41931.92593    35186.0  31639.50203   
2               459           378  84942.68317    74964.0  56811.62186   
3               438           358  48733.67116    37845.0  45100.54010   
4              1725          1540  31834.15466    22497.0  34046.50907   

   hi_sample_weight  hi_samples  family_mean  family_median  family_stdev  \
0        1290.96240      2024.0  67994.14790        53245.0   47667.30119   
1         838.74664      1127.0  50670.10337        43023.0   34715.57548   
2        1155.20980      2488.0  95262.51431        85395.0   49292.67664   
3         928.32193      1267.0  56401.68133        44399.0   41082.90515   
4        1548.67477      1983.0  54053.42396        50272.0   39609.12605   

   family_sample_weight  family_samples  hc_mortgage_mean  hc_mortgage_median  \
0             884.33516          1491.0        1414.80295              1223.0   
1             375.28798           554.0         864.41390               784.0   
2             709.74925          1889.0        1506.06758              1361.0   
3             490.18479           729.0        1175.28642              1101.0   
4             244.08903           395.0        1192.58759              1125.0   

   hc_mortgage_stdev  hc_mortgage_sample_weight  hc_mortgage_samples  \
0          641.22898                  377.83135                867.0   
1          482.27020                  316.88320                356.0   
2          731.89394                  699.41354               1491.0   
3 

## lets find the maximum value of male mean age and female mean age in the dataset.

In [55]:
train_data["male_age_mean"].max()

77.75992

In [57]:
train_data["female_age_mean"].max()

79.83739

## lets find the minimum value of male mean age and female mean age in the dataset.

In [61]:
train_data["male_age_mean"].min()

12.14583

In [59]:
train_data["female_age_mean"].min()

16.008329999999997

In [62]:
age_df = train_data[['state', 'city', 'place', 'pop', 'male_pop', 'female_pop', 'male_age_median', 'female_age_median', 'married', 'separated', 'divorced']]

## THE BRACKETS :

YOUNG YOUTH - <18 </br>
YOUTH - 19 to 30  </br>
ADULT - 30 to 55  </br>
SENIOR - >55      </br>

In [64]:
bins = [0,18, 30, 55, 100]
labels = ['YOUNG YOUTH', 'YOUTH', ' ADULLT' , 'SENIOR']

In [65]:
age_df['male_population_bracket'] = pd.cut(age_df['male_age_median'], bins, labels = labels)

In [66]:

age_df['female_population_bracket'] = pd.cut(age_df['female_age_median'], bins, labels = labels)

In [67]:
age_df.head()


state        city           place   pop  male_pop  female_pop  \
0     New York    Hamilton        Hamilton  5230      2612        2618   
1      Indiana  South Bend        Roseland  2633      1349        1284   
2      Indiana    Danville        Danville  6881      3643        3238   
3  Puerto Rico    San Juan        Guaynabo  2700      1141        1559   
4       Kansas   Manhattan  Manhattan City  5637      2586        3051   

   male_age_median  female_age_median  married  separated  divorced  \
0         44.00000           45.33333  0.57851    0.01240   0.08770   
1         32.00000           37.58333  0.34886    0.01426   0.09030   
2         40.83333           42.83333  0.64745    0.01607   0.10657   
3         48.91667           50.58333  0.47257    0.02021   0.10106   
4         22.41667           21.58333  0.12356    0.00000   0.03109   

  male_population_bracket female_population_bracket  
0                   Adult                     Adult  
1                   Adult                     Adult  
2                   Adult                     Adult  
3                   Adult                     Adult  
4                   YOUTH                     YOUTH

Load the df dataset in the excel sheet to perform visualisations.

In [71]:
age_df.to_excel('B.xlsx')

In [70]:
#--> USED POWERBI For visualisations